source/server/listener_impl.cc

```diff
@@ -221,6 +224,84 @@ Server::DrainManager& ListenerFactoryContextBaseImpl::drainManager() { return *d
 // Must be overridden
 Init::Manager& ListenerFactoryContextBaseImpl::initManager() { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
 
+PerFilterChainRebuilder::PerFilterChainRebuilder(
+    ListenerImpl& listener, const envoy::config::listener::v3::FilterChain* const& filter_chain,
+    Configuration::FactoryContext& factory_context)
+    : listener_(listener), filter_chain_(filter_chain), parent_context_(factory_context),
+      state_(State::Running),
+      rebuild_init_manager_(std::make_unique<Init::ManagerImpl>("rebuild_init_manager")),
+      rebuild_watcher_(
+          "rebuild_watcher",
+          [this] {
+            if (state_ == State::Running) {
+              ENVOY_LOG(debug, "rebuild_watcher is ready, rebuilder will callback to workers");
+              state_ = State::Succeeded;
+              callbackToWorkers(true);
+              if (timeout_enabled_) {
+                rebuild_timer_->disableTimer();
+              }
+            }
+          }),
+      rebuild_timer_(listener_.dispatcher().createTimer([this]() -> void { onTimeout(); })),
+      rebuild_timeout_(PROTOBUF_GET_MS_OR_DEFAULT(filter_chain->on_demand_configuration(),
+                                                  rebuild_timeout, 15000)) {}
+
+void PerFilterChainRebuilder::storeWorkerInCallbackList(const std::string& worker_name) {
+  workers_to_callback_.insert(worker_name);
+}
+
+Configuration::FilterChainFactoryContextPtr
+PerFilterChainRebuilder::createFilterChainFactoryContext(
+    const ::envoy::config::listener::v3::FilterChain* const filter_chain) {
+  UNREFERENCED_PARAMETER(filter_chain);
+  return std::make_unique<PerFilterChainFactoryContextImpl>(parent_context_, initManager());
+}
+
+void PerFilterChainRebuilder::callbackToWorkers(bool success) {
+  // Send callbacks to all stored workers.
+  // Possible optimization: send callback to all workers.
+  for (const auto& worker_name : workers_to_callback_) {
+    ENVOY_LOG(debug, "rebuilding completed, callback to worker: {}", worker_name);
+    if (listener_.hasWorker(worker_name)) {
+      auto& worker = listener_.getWorkerByName(worker_name);
+      worker->onFilterChainRebuilt(success, filter_chain_);
```

Comment:
The name to worker map is stable. The lookup could be avoided. Could we use worker pointer or worker reference.

Reply:
Updated